lightning/src/offers/merkle.rs

```diff
@@ -24,6 +25,50 @@ tlv_stream!(SignatureTlvStream, SignatureTlvStreamRef, SIGNATURE_TYPES, {
 	(240, signature: Signature),
 });
 
+/// A function for signing a [`Message`] digest produced from the given [`TaggedHash`]. See
+/// [`TaggedHash`] for details.
+///
+/// The second parameter is either the payer or offer metadata, depending on the TLV stream
+/// represented by the first parameter, and may be useful in order to derive the signing keys.
+pub trait SignFunction<E>: FnOnce(&TaggedHash, &[u8]) -> Result<Signature, E> {}
+
+impl<F, E> SignFunction<E> for F where F: FnOnce(&TaggedHash, &[u8]) -> Result<Signature, E> {}
+
+/// A hash for use in a specific context by tweaking with a context-dependent tag as per [BIP 340]
+/// and computed over the merkle root of a TLV stream to sign as defined in [BOLT 12].
+///
+/// The hash is computed lazily from the TLV stream.
+///
+/// [BIP 340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+/// [BOLT 12]: https://github.com/rustyrussell/lightning-rfc/blob/guilt/offers/12-offer-encoding.md#signature-calculation
+/// Bytes associated with a tag, which are used to produced a [`Message`] digest to sign.
+pub struct TaggedHash<'a> {
```

Comment:
nit: it's slightly unintuitive that this is called a "hash", but it actually contains the full stream. and in fact, the signer would look at the stream to validate.  perhaps it makes sense to call it something like `TaggedHashTlvStream` or `TlvStreamWithTaggedHash`

Reply:
Out of curiosity, how would the signer validate the stream? Presumably they may want to parse the message and do some checks. But when computing the hash, would calling `TaggedHash::to_message` be sufficient given this would be running on their process? Or would the signer not trust this code and opt to reimplement the hash computation?

Reply:
I guess you mean `to_digest`.  no need to reimplement anything, a code review is enough to trust code in dependencies, so there would be no problem calling that code.

if you are asking what further validation rules we could have on the actual data - we could have some sanity checks on the invoice fields (e.g. maximum amount).  but we would also like to handle the payment preimages in the future, so that an attacker can't redirect incoming payments to an intermediate hop.

Reply:
> I guess you mean `to_digest`. no need to reimplement anything, a code review is enough to trust code in dependencies, so there would be no problem calling that code.

Alright, so my argument for the name is that it's the input to the signing function, and the hash is what is being signed. The fact that the bytes that the hash were computed from is needed for some users is secondary.

> if you are asking what further validation rules we could have on the actual data - we could have some sanity checks on the invoice fields (e.g. maximum amount). but we would also like to handle the payment preimages in the future, so that an attacker can't redirect incoming payments to an intermediate hop. so we would check that the payment hash in the invoice is for a preimage we control

Hmm... so to reiterate what I said [earlier](https://github.com/lightningdevkit/rust-lightning/pull/2432#discussion_r1273994082), `NodeSigner::sign_bolt12_message` isn't actually called in the normal, privacy-preserving case. There's essentially three ways we allow creating an `Offer`:
- `OfferBuilder::new` which let's the user set any `metadata`,
- `OfferBuilder::deriving_signing_pubkey` without calling `OfferBuilder::path`, and
- `OfferBuilder::deriving_signing_pubkey` with calling `OfferBuilder::path` to add blinded paths.

Any `InvoiceRequest` messages created in the first case will be rejected by the `ChannelManager`.

The _last two cases_ will set `Offer::metadata` such that the `ChannelManager` implementation of `OffersMessageHandler` can verify that any `InvoiceRequest` came from one of our `Offer`s before replying with an `Invoice`.

Further, the _last case_ will set `Offer::signing_pubkey` to an ephemeral pubkey and allow `ChannelManager` to derive the corresponding secret key from the `Offer::metadata`. Thus, for this way, `NodeSigner::sign_bolt12_message` is not even called. Instead, `ChannelManager` will sign the invoice by deriving the keys from the metadata.

It's only the second case where `NodeSigner::sign_bolt12_message` is invoked. See PR #2371 for details. So it seems you'll need some other way to validate this data. Either by implementing `OffersMessageHandler` yourself or having `ChannelManager` call some other method on `NodeSigner` specifically for validation.

Reply:
> Further, the _last case_ will set `Offer::signing_pubkey` to an ephemeral pubkey and allow `ChannelManager` to derive the corresponding secret key from the `Offer::metadata`. Thus, for this way, `NodeSigner::sign_bolt12_message` is not even called. Instead, `ChannelManager` will sign the invoice by deriving the keys from the metadata.

Sorry for the slow reply and not fully understanding your first comment.  But I must be still missing something, since this seems potentially problematic.  How will a relying party know that this offer was securely issued by our node if the offer is not bound to our node key?  or looking at it from a different direction, why would something outside of the signer be able to enter into an agreement that could in general cause financial loss?

It's a design goal of the signer abstraction that nothing outside the signer can lose money, either by moving assets or by signing bogus invoices and such.

Reply:
> Sorry for the slow reply and not fully understanding your first comment. But I must be still missing something, since this seems potentially problematic. How will a relying party know that this offer was securely issued by our node if the offer is not bound to our node key?

The `signing_pubkey` is derived from `NodeSigner::get_inbound_payment_key_material`. Perhaps that should be renamed since it is not exclusively for inbound payments, though it is for this case. `ChannelManager` calls this upon initialization and uses the resulting `ExpandedKey` to verify that an `InvoiceRequest` originated from a known `Offer` using the included metadata.

Is that sufficient, or does the `NodeSigner` need to call `verify` directly?

> or looking at it from a different direction, why would something outside of the signer be able to enter into an agreement that could in general cause financial loss?
> 
> It's a design goal of the signer abstraction that nothing outside the signer can lose money, either by moving assets or by signing bogus invoices and such.

Here it's used for issuing invoices for _inbound_ payments, though it is also used for payer metadata in an `InvoiceRequest` and then verified when receiving an `Invoice` (i.e., outbound payments). For the latter case, the same derivation and verification as above applies.

Reply:
> Talking more abstractly about security model for VLS in BOLT12:
> 
> * For outbound payments, I feel like VLS should verify the HTLC we're sending by just receiving an `invoice` which it can detect matches up with a user-provided `offer` (and potentially amount) which it has listed as to-be-paid. This requires ~nothing from LDK aside from us giving them the `invoice` once we get it before sending the HTLC, and notably means VLS won't be involved in our metadata or `invoice_request` signing at all.

Do we want that in an interface separate from `NodeSigner`?

> * For inbound payments/`invoice` signing, VLS needs to be more involved as it wants to ensure we aren't signing invoices with bogus amounts for the expected payment, at least for those doing the non-private payments, as Jeff notes.

And for private payments, is the current flow ok without calling `NodeSigner`?

> Here too, I don't think VLS needs to be involved in the metadata (modern equivalent of payment_secret) generation/validation. However, in the non-private offer case, VLS/the signer wants to verify the BOLT12 offer/invoice data, at least the amount and description fields.

What's the reason why they would need to verify the data here but not for the private case?

> In that case, I don't think it makes sense to make the API read like you're only blindly signing a hash, but rather make it clear that there's data there you may want to validate.

Fair point. Is there any reason why `NodeSigner` needs to take the bytes directly? That is, can we simply give it the `Invoice` to be signed? Then `Invoice` can have methods for getting the underlying bytes or the `TaggedHash`, which it can also cache.

I may have used bytes initially to match `NodeSigner::sign_invoice` for BOLT 11 invoices. But the latter may be an artifact of not being able depend on `lightning-invoice` from `lightning`.

Reply:
This sounds right.  Here are some thoughts:

Issuing an invoice, `invoice_node_id == node_id`:

- the signer should be involved, because the payer relies on these and in general they are based on the `node_id`, which the signer controls

Issuing an invoice, `invoice_node_id != node_id`:

- not clear how this can be validated / controlled by the signer.  not clear how such a use-case would work.

Paying an invoice:

- the signer should be involved because the invoice justifies an outgoing HTLC.  the destination and amount are something that the signer should validate.  the signer should be presented with the invoice as a pre-flight before the HTLC is added (this is already the case for VLS with BOLT-11 invoicing)

Issuing an offer:

- it's not essential for the signer to control these, because an attacker will fail at the invoice issuance step.  also, these are unsigned, so no control is possible. 

Issuing an invoice request:

- cannot be controlled because `invreq_payer_id` is ephemeral.  seems like it's not important to control.

Reply:
> Indeed, this can't really include the signer as the device displaying the QR code can always replcae the offer, but we anticipate this being the "common" case - in most cases where a user is paying they're not validating the node_id in any way, so we should use an ephemeral key.

why do we think this is a common case?  if, for example, a user wants to pay some bill or buy some merchandise online, they would know what the node_id is (and hopefully allowlist the node_id to reduce having to approve things all the time).

> I kinda assumed we'd just have some call on the signer to say "hey, here's an invoice, we're paying it, FYI" and be done with it.

right, this is how we're handling bolt-11 invoices - we want to see the invoice before the HTLC happens, and check allowlist / velocity / user approval via secure UX, ahead of the payment starting it's flight.

that said, I'm not sure how useful the *offer* is, given that it's unsigned and could be arbitrarily generated by anybody.

Reply:
I see.  in this case, we should have the approval UX pre-approve an offer for repeated payments.  so we should use either the node_id or the offer_id as the "destination".

Reply:
Note, the spec no longer has an `offer_id`.

Reply:
> I kinda assumed we'd just have some call on the signer to say "hey, here's an invoice, we're paying it, FYI" and be done with it. Presumably any VLS client will pass VLS any offers which the client has authorized to pay via some secure channel (presumably also how it works today with BOLT11) or has some policy enforcement. So all LDK really needs to do is give VLS a mapping from the offer to the invoice that we're about to pay, and VLS can validate that the invoice is built using the offer.

Can't VLS use the metadata to verify an invoice instead of needing the client to pass it authorized offers? Presumably, it has the same key material as the `NodeSigner` and thus can form an `ExpandedKey` to verify. Or am I misunderstanding the setup.